ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPartyManagementService.scala

```diff
@@ -47,24 +48,57 @@ class ApiPartyManagementService private (
       .listParties()
       .map(ps => ListKnownPartiesResponse(ps.map(mapPartyDetails)))(DE)
 
+  /**
+    * Continuously polls tha party management service to check if a party has been persisted.
+    *
+    * Blocking.
+    *
+    * @param party The party whose persistence we're waiting for
+    * @return The number of attempts before the party was found
+    */
+  private def pollUntilPersisted(party: String): Int =
+    Iterator
+      .from(1)
+      .find { attempt =>
+        logger.debug(s"Waiting for party '$party' to be persisted (attempt #$attempt)...")
+        Await.result(
+          partyManagementService
+            .listParties()
+            .map {
+              case persisted if persisted.exists(_.party == party) => true
+              case _ => false
+            }(DE),
+          Duration.Inf)
+      }
+      .get
+
   override def allocateParty(request: AllocatePartyRequest): Future[AllocatePartyResponse] = {
     val party = if (request.partyIdHint.isEmpty) None else Some(request.partyIdHint)
     val displayName = if (request.displayName.isEmpty) None else Some(request.displayName)
 
-    FutureConverters
-      .toScala(
-        writeService
-          .allocateParty(party, displayName))
-      .flatMap {
-        case PartyAllocationResult.Ok(details) =>
-          Future.successful(AllocatePartyResponse(Some(mapPartyDetails(details))))
-        case r @ PartyAllocationResult.AlreadyExists =>
-          Future.failed(ErrorFactories.invalidArgument(r.description))
-        case r @ PartyAllocationResult.InvalidName(_) =>
-          Future.failed(ErrorFactories.invalidArgument(r.description))
-        case r @ PartyAllocationResult.ParticipantNotAuthorized =>
-          Future.failed(ErrorFactories.permissionDenied(r.description))
-      }(DE)
+    val response =
+      FutureConverters
+        .toScala(
+          writeService
+            .allocateParty(party, displayName))
+        .flatMap {
+          case PartyAllocationResult.Ok(details) =>
+            Future.successful(AllocatePartyResponse(Some(mapPartyDetails(details))))
+          case r @ PartyAllocationResult.AlreadyExists =>
+            Future.failed(ErrorFactories.invalidArgument(r.description))
+          case r @ PartyAllocationResult.InvalidName(_) =>
+            Future.failed(ErrorFactories.invalidArgument(r.description))
+          case r @ PartyAllocationResult.ParticipantNotAuthorized =>
+            Future.failed(ErrorFactories.permissionDenied(r.description))
+        }(DE)
+
+    // Since we are using the direct execution context above, the future is actually already available here
```

Comment:
That's not quite true. Using the direct execution context means the action runs on the same thread that the future anyway is already running on. Using `Await.result` here means that we are blocking in the grpc thread until the work on the database thread is complete.

Reply:
I got it wrong, thank you for the explanation.